grails-app/jobs/org/pih/warehouse/jobs/RefreshInventorySnapshotAfterTransactionJob.groovy

```diff
@@ -0,0 +1,87 @@
+ * By using this software in any fashion, you are agreeing to be bound by
+ * the terms of this license.
+ * You must not remove this notice, or any other, from this software.
+ **/
+package org.pih.warehouse.jobs
+
+import org.codehaus.groovy.grails.commons.ConfigurationHolder
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.inventory.Transaction
+import org.pih.warehouse.product.Product
+import org.quartz.DisallowConcurrentExecution
+import org.quartz.JobExecutionContext
+import org.quartz.JobExecutionException
+
+@DisallowConcurrentExecution
+class RefreshInventorySnapshotAfterTransactionJob {
+
+    def concurrent = false
+    def grailsApplication
+    def inventorySnapshotService
+
+    static triggers = {}
+
+    def execute(JobExecutionContext context) {
+
+        Boolean enabled = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotAfterTransactionJob.enabled
+        log.info("Refresh inventory snapshots with data (enabled=${enabled}): " + context.mergedJobDataMap)
+        if (enabled) {
+
+            def jobDataMap = context.jobDetail.jobDataMap
+            Integer retryCount = jobDataMap.containsKey("retryCount") ? jobDataMap.getIntFromString("retryCount"):0
+            Integer maxRetryAttempts = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotJob.maxRetryAttempts?:3
+            Boolean retryOnError = ConfigurationHolder.config.openboxes.jobs.refreshInventorySnapshotJob.retryOnError?:false
+            if (retryOnError) {
+                log.info "Retry count: ${retryCount} / ${maxRetryAttempts}"
+                if (retryOnError && retryCount >= maxRetryAttempts) {
+                    JobExecutionException e = new JobExecutionException("Retries exceeded")
+                    e.setUnscheduleAllTriggers(true)
+                    throw e
+                }
+            }
+
+            boolean forceRefresh = context.mergedJobDataMap.getBoolean("forceRefresh")
+            boolean isDeleting = context.mergedJobDataMap.getBoolean("isDeleting")
+            def startTime = System.currentTimeMillis()
+            def userId = context.mergedJobDataMap.get('user')
+            def startDate = context.mergedJobDataMap.get('startDate')
+            String locationId = context.mergedJobDataMap.get('location')
+            def productIds = context.mergedJobDataMap.get('productIds')
+            Location location = Location.get(locationId)
+
+            try {
+                log.info ("Refresh inventory snapshot " + retryCount + " out of " + maxRetryAttempts)
+
+                def date = new Date() + 1
+
+                productIds.each { String productId ->
+                    Product product = Product.get(productId)
+
+                    if (forceRefresh) {
+                        inventorySnapshotService.deleteInventorySnapshots(location, product)
+                    }
+
+                    if (!isDeleting) {
```

Comment:
Can you explain this? I think we still want to refresh the inventory snapshots even when we're deleting a transaction.

Reply:
@jmiranda for example:
I had snapshot of the inventory of two products with some quantity. After deleting the transaction for these two entries they were readded to the inventory snapshot table during populateInventorySnapshots (with the same qoh values as it was before deleting). Is this desired behavior? Is there something that indicates if the inventory was deleted even if the quantity on hand has positive value?

Reply:
I'll reinvestigate it tomorrow together with the comment related to pulling associatedProducts from transaction.

Reply:
So let's say we have a Product A and that product has two items (A123 and A456). 

State A
Initial inventory is 0.

State B
We Record Stock and quantity for A123 is 10 and A456 is 10. Product A QoH = 20. Inventory Snapshot process runs and saves the two rows.

State C
Then we create a DEBIT transaction with quanitty for A123 = 10. So now we only A456 = 10. Product A = 10. Inventory Snapshot process runs and saves the one row.

State D
Then we delete the DEBIT transaction which reverts the state to the State B. The inventory snapshot process should run and save two records to the table. Product A QoH = 20.

Reply:
Make sure that works properly and document any other issues (like the race condition) in the ticket.